### ISSUE TITLE
Make `Projector` instances pickleable

### ISSUE THREAD

**Issue by araghukas** [PennyLaneAI/pennylane#4443]
### Feature details

Instances of `qml.Projector` can't be pickled at the moment:
```python
import pickle
import pennylane as qml

P = qml.Projector([1, 0, 1, 1, 0], wires=[0, 1, 2, 3, 4])   # specify w/ basis state
pickle.dumps(P)
```

```
_pickle.PicklingError: Can't pickle <class 'pennylane.ops.qubit.observables.Projector'>: it's not the same object as pennylane.ops.qubit.observables.Projector
```
It would be nice for `qml.Projector` instances to pickle/unpickle normally again (as prior to v0.31). 

### Implementation

This PR (https://github.com/PennyLaneAI/pennylane/pull/3588) is very helpful.

#### Possible Explanation

Based on the `PicklingError` above, the instance being pickled has `type(P)` which is not identical to `Projector` as imported from `pennylane.ops.qubit.observables`. The [docstring](https://github.com/PennyLaneAI/pennylane/blob/26c2655c2001dd0d1defabc151fb649795750cb0/pennylane/ops/qubit/observables.py#L377) of `Projector.__new__()` hints at the reason:
 ```
        Though all the types will be named "Projector", their *identity* and location in memory
        will be different based on whether the input state is a basis state or a state vector.
```
We can see from the source that an "undifferentiated" `Projector` type is [modified during instantiation](https://github.com/PennyLaneAI/pennylane/blob/26c2655c2001dd0d1defabc151fb649795750cb0/pennylane/ops/qubit/observables.py#L398) (before `__init__()`). So the undifferentiated class becomes a *parent class* for the instance. Note there are two "Projector" classes in the following.
```python
print(P.__class__.mro())
```
```
[<class 'pennylane.ops.qubit.observables.Projector'>,
<class 'pennylane.ops.qubit.observables._BasisStateProjector'>,
<class 'pennylane.ops.qubit.observables.Projector'>, 
<class 'pennylane.operation.Observable'>, 
<class 'pennylane.operation.Operator'>, 
<class 'abc.ABC'>, 
<class 'object'>]
```

The issue is that `pickle` requires the imported (undifferentiated) type to be the same type as the instance's. But this is not the case with `Projector`:
```python
import pennylane as qml

P = qml.Projector([1, 0, 1, 1, 0], wires=[0, 1, 2, 3, 4])

print(isinstance(P, qml.Projector))
print(type(P) is qml.Projector)
```
```
True
False
```

#### Suggestion

To avoid the `PicklingError`, we can explicitly provide a type to consider when pickling.

Adding a [`__reduce__()`](https://docs.python.org/3/library/pickle.html#object.__reduce__) method to `Projector` serves exactly this purpose:
```python
    def __reduce__(self):
        """Defines how to pickle and unpickle a Projector. This is needed
        because new instances have a dynamic type that inherits from Projector,
        but is not identical to Projector."""
        state = self.data[0]
        return (Projector, (state, self.wires))
```

### How important would you say this feature is?

1: Not important. Would be nice to have.

### Additional information

Adding `__reduce__()` also enables pickling with the non-standard `cloudpickle` package.

**Comment by josh146** [comment]
Hi @araghukas! Thank you for this very detailed bug report 🙏 While the team looks into this, can I ask what workflow you were using that required Pickle? Was the pickling being done manually, or via a third party library (e.g., Ray, Dask, etc.).

**Comment by araghukas** [comment]
Thanks @josh146.

In my case the pickling for [Covalent](https://github.com/AgnostiqHQ/covalent). It uses pickle extensively to transport objects to/from remote backends.

**Comment by BorjaRequena** [comment]
Hi @araghukas, thanks for reporting this issue! I have put up a fix for this in #4452. There are a few details we need to polish, but it should be ready very soon! 🚀

**Comment by araghukas** [comment]
Great to hear! Thanks @BorjaRequena.
